### PR TITLE
ui Prevent tool messages from incorrectly appending to other conversations

### LINKED ISSUE #24337: Misc. bug:  Content from active agentic loop is invading another chats
### Name and Version

While the streaming is active in one chat, when we switch to another one, the agentic loop is suddelny showing up in another chats.

### Operating systems

_No response_

### Which llama.cpp modules do you know to be affected?

_No response_

### Command line

```shell

```

### Problem description & steps to reproduce

- Start agentic loop in one convo
- Switch to another convo
- See that the 1st convo's generated content suddenly is being rendered in the 2nd convo
- After the generation finishes and we refresh, the convos are proper, i. e. the content that invaded the 2nd convo is not there and is only in the 1st

### First Bad Commit

_No response_

### Relevant log output

<details>
<summary>Logs</summary>

```console

```
</details>

### PR DESCRIPTION
**Pull request by allozaur**
Close #24337

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES, pi with Qwen3.6-35B-A3B

### CODE REVIEW

**Comment by ServeurpersoCom** [comment]
That’s a really smart point. I’ve never switched conversations during an agentic loop. It seems to me that the bot is right about the edge case. I'll test it out in practice.

**Comment by ServeurpersoCom** [comment]
Here is the reproduction case, it's a bit tricky.

It's a race during the MCP response: ask the model to generate 10 "sleep 1" toolcalls, one at a time, sequentially.

While the agentic loop is running, edit a previous message in a different conversation, which creates multiple versions of that edited message. That's where it breaks (it jump to the first edited version of the message). Here is the fix, tested on my side :

```
root@pod:/mnt/workspace/git/llama.cpp# git diff
diff --git a/tools/ui/src/lib/stores/chat.svelte.ts b/tools/ui/src/lib/stores/chat.svelte.ts
index 873f960bfd..6949047b50 100644
--- a/tools/ui/src/lib/stores/chat.svelte.ts
+++ b/tools/ui/src/lib/stores/chat.svelte.ts
@@ -1213,8 +1213,15 @@ class ChatStore {
                                };
                                if (timings) uiUpdate.timings = timings;
                                if (resolvedModel) uiUpdate.model = resolvedModel;
-                               conversationsStore.updateMessageAtIndex(idx, uiUpdate);
-                               await conversationsStore.updateCurrentNode(currentMessageId);
+                               // touch the active ui array and node pointer only when this conversation
+                               // is displayed; otherwise persist the node move straight to the db so a
+                               // foreign conv's currNode stays untouched
+                               if (conversationsStore.activeConversation?.id === convId) {
+                                       conversationsStore.updateMessageAtIndex(idx, uiUpdate);
+                                       await conversationsStore.updateCurrentNode(currentMessageId);
+                               } else {
+                                       await DatabaseService.updateCurrentNode(convId, currentMessageId);
+                               }
                        },
                        createToolResultMessage: async (
                                toolCallId: string,
@@ -1235,12 +1242,15 @@ class ChatStore {
                                        },
                                        currentMessageId
                                );
-                               // mirror into the active store only when this conversation is displayed;
-                               // otherwise the tool message would land in another conv's UI array
+                               // mirror into the active store and move the node pointer only when this
+                               // conversation is displayed; otherwise persist the node move straight to
+                               // the db for the owning conv so a foreign conv's currNode stays untouched
                                if (conversationsStore.activeConversation?.id === convId) {
                                        conversationsStore.addMessageToActive(msg);
+                                       await conversationsStore.updateCurrentNode(msg.id);
+                               } else {
+                                       await DatabaseService.updateCurrentNode(convId, msg.id);
                                }
-                               await conversationsStore.updateCurrentNode(msg.id);
                                lastCreatedInFlow = msg.id;
                                return msg;
                        },
root@pod:/mnt/workspace/git/llama.cpp#
```

It's okay for PR with this.

**Review by ServeurpersoCom — APPROVED** [review]
LGTM with the provided patch. rabbit missed the twin in onAssistantTurnComplete, same ungated updateCurrentNode, needs the same gating or the jump still happens on turn completion.

**Comment by ServeurpersoCom** [comment]
Oh darn, there's one more commit needed to make it perfect.

**Comment by ServeurpersoCom** [comment]
Without fix :

https://github.com/user-attachments/assets/6a66b7b6-e822-4304-8e7c-7f56b5bfa506

Half-fixed :

https://github.com/user-attachments/assets/3eea2a89-fdef-44c8-aed5-91e7fadcc5cc

**Comment by ServeurpersoCom** [comment]
Mini-commit covering the final case: I replaced $derived with $state for editedContent to prevent a background agentic recompute from wiping the in-progress edit buffer.

Fully fixed :

https://github.com/user-attachments/assets/807467e8-976e-4394-86f0-d631be439d85

**Comment by allozaur** [comment]
Thanks for contributing to this fix @ServeurpersoCom :)

**Comment by allozaur** [comment]
@ggml-org/maintainers can we get a 2nd approval for this fix?

**Review by 0cc4m — APPROVED** [review]
